tool/smarts/src/main/java/org/openscience/cdk/smarts/SmartsFragmentExtractor.java

```diff
@@ -0,0 +1,455 @@
+/*
+ * Copyright (c) 2016 John May <jwmay@users.sf.net>
+ *
+ * Contact: cdk-devel@lists.sourceforge.net
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or (at
+ * your option) any later version. All we ask is that proper credit is given
+ * for our work, which includes - but is not limited to - adding the above
+ * copyright notice to the beginning of your source code files, and to any
+ * copyright notice that you may distribute with programs based on this work.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public
+ * License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 U
+ */
+
+package org.openscience.cdk.smarts;
+
+import org.openscience.cdk.interfaces.IAtom;
+import org.openscience.cdk.interfaces.IAtomContainer;
+import org.openscience.cdk.interfaces.IBond;
+
+import java.util.Arrays;
+import java.util.Locale;
+
+/**
+ * Utility class to create SMARTS that match part (substructure) of a molecule.
+ * SMARTS are generated by providing the atom indexes. An example use cases is
+ * encoding features from a fingerprint.
+ * <p>
+ * The extractor has two modes. {@link #MODE_EXACT} (default) captures the element,
+ * valence, hydrogen count, connectivity, and charge in the SMARTS atom expressions.
+ * The alternative mode, {@link #MODE_JCOMPOUNDMAPPER}, only captures the element,
+ * non-zero charge, and peripheral bonds. Although the later looks cleaner, the
+ * peripheral bonds intend to capture the connectivity of the terminal atoms but
+ * since the valence is not bounded further substitution is still allowed.
+ *
+ * <p>The difference is easily demonstrated for methyl. Consider the compound
+ * of 2-methylpentane {@code CC(C)CCC}, if we extract one of the methyl atoms
+ * depending on the mode we obtain {@code [CH3v4X4+0]} or {@code C*}. The first
+ * of these patterns (obtained by {@link #MODE_EXACT}) matches the compound in
+ * <b>three places</b> (the three methyl groups). The second matches <b>six</b>
+ * times (every atom) because the substituion on the carbon is not locked.
+ * A further complication is introduced by the inclusion of the peripheral atoms,
+ * for 1H-indole {@code [nH]1ccc2c1cccc2} we can obtain the SMARTS {@code n(ccc(a)a)a}
+ * that doesn't match at all. This is because one of the aromatic atoms ('a')
+ * needs to match the nitrogen.
+ *
+ * <p><b>Basic Usage:</b></p>
+ * <pre>{@code
```

Comment:
What is "{@code"? Should we start using that for all JavaDoc? If so, sounds like a nice Junior Job...